File: Java/OPSJLib/src/ops/Participant.java
/**
 *
 * Copyright (C) 2006-2009 Anton Gravestam.
 *
 * This file is part of OPS (Open Publish Subscribe).
 *
 * OPS (Open Publish Subscribe) is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.

 * OPS (Open Publish Subscribe) is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with OPS (Open Publish Subscribe).  If not, see <http://www.gnu.org/licenses/>.
 */
package ops;

import java.lang.management.ManagementFactory;
import java.util.logging.Level;
import java.util.logging.Logger;
import configlib.SerializableFactory;
import configlib.exception.FormatException;
import java.io.File;
import java.io.IOException;
import ops.archiver.OPSObjectFactory;
import ops.transport.inprocess.InProcessTransport;
import ops.ConfigurationException;

/**
 *
 * @author Anton Gravestam
 */
public class Participant
{
    //This factory is used through static facade methods getInstance(...)
    private static ParticipantFactory participantFactory = new ParticipantFactory();

    protected String domainID;
    private String participantID;
    protected Domain domain;
    private ErrorService errorService = new ErrorService();
    private ReceiveDataHandlerFactory receiveDataHandlerFactory = new ReceiveDataHandlerFactory();
    private SendDataHandlerFactory sendDataHandlerFactory = new SendDataHandlerFactory();
    private InProcessTransport inProcessTransport = new InProcessTransport();

    private final ParticipantInfoData partInfoData = new ParticipantInfoData();
    private Publisher partInfoPub = null;
    private boolean keepRunning = false;

    ///LA Added
    private OPSConfig config = null;

    /**
     * Method for retreiving the default Participant instance for the @param domainID
     * @param domainID
     * @return a Participant or null if this method fails.
     */
    public static synchronized Participant getInstance(String domainID) throws ConfigurationException
    {
        return participantFactory.getParticipant(domainID, "DEFAULT_PARTICIPANT");
    }

    /**
     * Method for retreiving the @param participantID Participant instance for @param domainID
     * @param domainID
     * @param participantID
     * @return a Participant or null if this method fails.
     */
    public static synchronized Participant getInstance(String domainID, String participantID) throws ConfigurationException
    {
        return participantFactory.getParticipant(domainID, participantID, null);
    }

    /**
     * Method for retreiving the @param participantID Participant instance for
     * @param domainID using @param file if this participant has not yet been created.
     * @param domainID
     * @param participantID
     * @param file the file to use as configuration. Ignored if a participant for
     * participantID already exist.
     * @return a Participant or null if this method fails.
     */
    public static synchronized Participant getInstance(String domainID, String participantID, File file) throws ConfigurationException
    {
        return participantFactory.getParticipant(domainID, participantID, file);
    }


    public static synchronized Participant getInstance(Domain domain, String participantID)
    {
        return participantFactory.getParticipant(domain, participantID);
    }

    protected Participant(String domainID, String participantID, File configFile) throws ConfigurationException
    {
        this.domainID = domainID;
        this.participantID = participantID;
        try
        {
            if (configFile == null)
            {
                config = OPSConfig.getConfig();
                domain = config.getDomain(domainID);
            }
            else
            {
                config = OPSConfig.getConfig(configFile);
                domain = config.getDomain(domainID);
            }
            if (domain != null)
            {
                inProcessTransport.start();
                setupCyclicThread();
            }
            else
            {
                throw new ConfigurationException("Participant(): Failed to find requested domain in configuration file");
            }
        }
        catch (IOException ex)
        {
            throw new ConfigurationException("Participant(): Configuration file missing");
        }
        catch (FormatException ex)
        {
            throw new ConfigurationException("Participant(): Format error in Configuration file");
        }
    }

    protected Participant(Domain domain, String participantID)
    {
        this.domainID = domain.getDomainID();
        this.participantID = participantID;
        
        this.domain = domain;

        inProcessTransport.start();
        setupCyclicThread();
    }

    ///LA Added
    /**
     * Return the OPSConfig object used (if any).
     * Note that null can be returned if Participant is created from a Domain object
     */
    public OPSConfig getConfig()
    {
        return config;
    }

    // Initialize static data in partInfoData
    private void initPartInfoData()
    {
        String processId = "";
        String localhostname = "";
        try
        {
            processId = ManagementFactory.getRuntimeMXBean().getName().split("@")[0];
            localhostname = java.net.InetAddress.getLocalHost().getHostName();
        } catch (Exception e)
        {
        }
        String Name = localhostname + " (" + processId + ")";

        synchronized (partInfoData)
        {
            partInfoData.name = Name;
            partInfoData.languageImplementation = "Java";
            partInfoData.id = participantID;
            partInfoData.domain = domainID;
            partInfoData.ip = "";
            partInfoData.opsVersion = "";
        }
    }

    public void setUdpTransportInfo(String ip, int port)
    {
        synchronized (partInfoData)
        {
            partInfoData.ip = ip;
            partInfoData.mc_udp_port = port;
        }
    }

    //
    public boolean hasPublisherOn(String topicName)
    {
        return true;    ///TODO
    }

    /**
     * Remove listener for Error events.
     * @param listener
     */
    public void removeListener(Listener<Error> listener)
    {
        errorService.removeListener(listener);
    }

    /**
     * Add listener for Error events.
     * @param listener
     */
    public synchronized void addListener(Listener<Error> listener)
    {
        errorService.addListener(listener);
    }

    /**
     * Report a core Error.
     * @param className the class in which the error occured.
     * @param methodName the method in which the error occured.
     * @param errorMessage a message descriing the error.
     */
    void report(String className, String methodName, String errorMessage)
    {
        errorService.report(className, methodName, errorMessage);
    }

    /**
     * Report a core Error
     * @param error the Error to report.
     */
    void report(Error error)
    {
        errorService.report(error);
    }

    
    /**
     * Adds type support in form of a SerializablFactory to this participant.
     * Normally this is the IDL project generated TypeFactory (e.g. FooProject.FooProjectTypeFactory())
     * @param typeSupport
     */
    public void addTypeSupport(SerializableFactory typeSupport)
    {
        OPSObjectFactory.getInstance().add(typeSupport);
    }

    /**
     * Creates a Topic that can be used to create Publishers and/or Subscribers.
     * The fields of the Topic returned are fetched from the participants underlying config file.
     * @param name
     * @return a new Topic based on the config of this participant.
     */
    public Topic createTopic(String name)
    {
        Topic topic = domain.getTopic(name);
        
        if(topic == null)
            return null;

        topic.setParticipantID(participantID);
        topic.setDomainID(domainID);

        return topic;
    }

    /**
     * Creates a Topic that can be used to Subscribe to the participant info data
     * @return a new Topic for the participant info data.
     */
    public Topic createParticipantInfoTopic()
    {
   	Topic infoTopic = new Topic("ops.bit.ParticipantInfoTopic", domain.GetMetaDataMcPort(), "ops.ParticipantInfoData", domain.getDomainAddress());
    	infoTopic.setDomainID(domainID);
        infoTopic.setParticipantID(participantID);
	infoTopic.setTransport(Topic.TRANSPORT_MC);
    	return infoTopic;
    }

    ///By modified singelton
    synchronized ReceiveDataHandler getReceiveDataHandler(Topic top)
    {
        ReceiveDataHandler rdh = receiveDataHandlerFactory.getReceiveDataHandler(top, this);
        if (rdh != null)
        {
            synchronized (partInfoData)
            {
                partInfoData.subscribeTopics.add(new TopicInfoData(top));
            }
        }
        return rdh;
    }

    synchronized void releaseReceiveDataHandler(Topic topic)
    {
        receiveDataHandlerFactory.ReleaseReceiveDataHandler(topic, this);

        synchronized (partInfoData)
        {
            for (int i = 0; i < partInfoData.subscribeTopics.size(); i++)
            {
    		if (partInfoData.subscribeTopics.elementAt(i).name.equals(topic.getName())) {
                    partInfoData.subscribeTopics.remove(i);
                    break;
		}
            }
        }
    }

    ///By modified singelton
    synchronized SendDataHandler getSendDataHandler(Topic t) throws CommException
    {
        return sendDataHandlerFactory.getSendDataHandler(t, this);
    }

    public Domain getDomain()
    {
        return domain;
    }

    public InProcessTransport getInProcessTransport()
    {
        return inProcessTransport;
    }

    private void setupCyclicThread()
    {
        keepRunning = true;
        Thread thread = new Thread(new Runnable()
        {
            public void run()
            {
                initPartInfoData();

                while (keepRunning)
                {
                    try
                    {
                        Thread.sleep(1000);

                        // Create publisher for participant info data
                        if ( (partInfoPub == null) && (domain.GetMetaDataMcPort() != 0) )
                        {
                            partInfoPub = new Publisher(createParticipantInfoTopic());
                        }

                        // Publish data
                        if (partInfoPub != null)
                        {
                            synchronized (partInfoData)
                            {
                                partInfoPub.writeAsOPSObject(partInfoData);
                            }
                        }
                    }
                    catch (InterruptedException ex)
                    {
                        Logger.getLogger(Participant.class.getName()).log(Level.SEVERE, null, ex);
                    }
                    catch (ConfigurationException ex)
                    {
                        Logger.getLogger(Participant.class.getName()).log(Level.SEVERE, null, ex);
                    }
                }
            }
        });

        thread.setName("ParticipantThread_" + domainID + "_" + participantID);
        thread.start();
    }

}

File: trunk/C#/Ops/Participant.cs
///////////////////////////////////////////////////////////
//  Participant.cs
//  Implementation of the Class Participant
//  Created on:      12-nov-2011 09:25:33
//  Author:
///////////////////////////////////////////////////////////

using System.IO;
using System.Net;
using System.Runtime.CompilerServices;  // Needed for the "MethodImpl" synchronization attribute
using System.Threading;

namespace Ops 
{
	public class Participant
    {
        // This factory is used through static facade methods getInstance(...)
        private static ParticipantFactory participantFactory = new ParticipantFactory();

        private OPSConfig config;
        protected Domain domain;
        public string domainID { get; private set; }
		private InProcessTransport inProcessTransport = new InProcessTransport();
        public string participantID { get; private set; }
		private ReceiveDataHandlerFactory receiveDataHandlerFactory = new ReceiveDataHandlerFactory();
		private SendDataHandlerFactory sendDataHandlerFactory = new SendDataHandlerFactory();

        ///The data type factory used in this Participant. 
        private OPSObjectFactory objectFactory = new OPSObjectFactory();

        private ParticipantInfoData partInfoData = new ParticipantInfoData();
        private Publisher partInfoPub = null;

        private Thread thread = null;
        private bool doRun = true;

        private static int safeInstanceCount = 0;

        static public int SafeInstanceCount
        {
            get { return safeInstanceCount; }
        }

        /**
         * Method for retreiving the default Participant instance for the @param domainID
         * @param domainID
         * @return a Participant or null if this method fails.
         */
        [MethodImpl(MethodImplOptions.Synchronized)]
        public static Participant GetInstance(string domainID)
        {
            return participantFactory.GetParticipant(domainID, "DEFAULT_PARTICIPANT");
        }

        /**
         * Method for retreiving the @param participantID Participant instance for @param domainID
         * @param domainID
         * @param participantID
         * @return a Participant or null if this method fails.
         */
        [MethodImpl(MethodImplOptions.Synchronized)]
        public static Participant GetInstance(string domainID, string participantID)
        {
            return participantFactory.GetParticipant(domainID, participantID, null);
        }

        /**
         * Method for retreiving the @param participantID Participant instance for
         * @param domainID using @param file if this participant has not yet been created.
         * @param domainID
         * @param participantID
         * @param file the file to use as configuration. Ignored if a participant for participantID already exist.
         * @return a Participant or null if this method fails.
         */
        [MethodImpl(MethodImplOptions.Synchronized)]
        public static Participant GetInstance(string domainID, string participantID, string configFile)
        {
            return participantFactory.GetParticipant(domainID, participantID, configFile);
        }

        [MethodImpl(MethodImplOptions.Synchronized)]
        public static Participant GetInstance(Domain domain, string participantID)
        {
            return participantFactory.GetParticipant(domain, participantID);
        }

        //---------------------------------------------------------------

        public Participant(string domainID, string participantID, string configFile)
        {
            Interlocked.Increment(ref safeInstanceCount);  ///TEST
            this.domainID = domainID;
            this.participantID = participantID;
            try
            {
                if (configFile == "")
                {
                    this.config = OPSConfig.GetConfig();
                    this.domain = this.config.GetDomain(domainID);
                } else
                {
                    this.config = OPSConfig.GetConfig(configFile);
                    this.domain = this.config.GetDomain(domainID);
                }
                if (this.domain != null)
                {
                    this.inProcessTransport.Start();
                    SetupCyclicThread();
                }
                else
                {
                    Logger.ExceptionLogger.LogMessage(this.GetType().Name + ", Failed to find requested domain in configuration file");
                    throw new System.Exception("Failed to find requested domain in configuration file");
                }
            }
            catch (IOException)
            {
                Logger.ExceptionLogger.LogMessage(this.GetType().Name + ", Configuration file missing");
                throw;
            }
            catch (Ops.FormatException)
            {
                Logger.ExceptionLogger.LogMessage(this.GetType().Name + ", Format error in configuration file");
                throw;
            }
        }

        public Participant(Domain domain, string participantID)
        {
            Interlocked.Increment(ref safeInstanceCount);  ///TEST
            this.domainID = domain.GetDomainID();
            this.participantID = participantID;
            
            this.domain = domain;

            this.inProcessTransport.Start();
            SetupCyclicThread();
        }

        ~Participant()
        {
            Interlocked.Decrement(ref safeInstanceCount);   ///TEST
        }

        /// <summary>
        /// Method used to Stop the participant and remove it from the participant factory.
        /// The participant object will be deleted when no one have references to it.
        /// </summary>
        [MethodImpl(MethodImplOptions.Synchronized)]
        public void StopAndReleaseResources()
        {
            // Stop the cyclic thread
            if (this.doRun)
            {
                this.doRun = false;
                this.thread.Join(3000); // Timeout just for safety
            }
            // Remove us from the singleton factory
            participantFactory.RemoveParticipant(this.domainID, this.participantID);
            sendDataHandlerFactory = null;
        }

        // Initialize static data in partInfoData
        private void InitPartInfoData()
        {
            string processId = System.Diagnostics.Process.GetCurrentProcess().Id.ToString();
            System.Net.NetworkInformation.IPGlobalProperties gp = System.Net.NetworkInformation.IPGlobalProperties.GetIPGlobalProperties();
            string Name = gp.HostName + " (" + processId + ")";

            lock (partInfoData)
            {
                partInfoData.name = Name;
                partInfoData.languageImplementation = "C#";
                partInfoData.id = participantID;
                partInfoData.domain = domainID;
            }
        }

        public void SetUdpTransportInfo(string ip, int port)
        {
            lock (partInfoData)
            {
                partInfoData.ip = ip;
                partInfoData.mc_udp_port = port;
            }
        }

        // 
        public bool HasPublisherOn(string topicName)
        {
            return true;    ///TODO
        }

        /**
         * Adds type support in form of a SerializablFactory to this participant.
         * Normally this is the IDL project generated TypeFactory (e.g. FooProject.FooProjectTypeFactory())
         * @param typeSupport
         */
        public void AddTypeSupport(ISerializableFactory typeSupport)
        {
            this.objectFactory.Add(typeSupport);
        }

        ///Get a pointer to the data type factory used in this Participant. 
        public OPSObjectFactory getObjectFactory()
        {
            return this.objectFactory;
        }

        /**
         * Creates a Topic that can be used to create Publishers and/or Subscribers.
         * The fields of the Topic returned are fetched from the participants underlying config file.
         * @param name
         * @return a new Topic based on the config of this participant.
         */
        public Topic CreateTopic(string name)
        {
            Topic topic = domain.GetTopic(name);
            
            if (topic == null)
                return null;

            topic.SetParticipantID(participantID);
            topic.SetDomainID(domainID);

            return topic;
        }

        /**
         * Creates a Topic that can be used to Subscribe to the participant info data
         * @return a new Topic for the participant info data.
         */
        public Topic CreateParticipantInfoTopic()
    	{
    		Topic infoTopic = new Topic("ops.bit.ParticipantInfoTopic", domain.GetMetaDataMcPort(), "ops.ParticipantInfoData", domain.GetDomainAddress());
	    	infoTopic.SetDomainID(domainID);
		    infoTopic.SetParticipantID(participantID);
    		infoTopic.SetTransport(Topic.TRANSPORT_MC);
	    	return infoTopic;
    	}

        // By modified singelton
        [MethodImpl(MethodImplOptions.Synchronized)]
        public ReceiveDataHandler GetReceiveDataHandler(Topic topic)
        {
            ReceiveDataHandler rdh = this.receiveDataHandlerFactory.GetReceiveDataHandler(topic, this);
            if (rdh != null)
            {
                lock (partInfoData)
                {
                    partInfoData.subscribeTopics.Add(new TopicInfoData(topic));
                }
            }
            return rdh;
        }

        [MethodImpl(MethodImplOptions.Synchronized)]
        public void ReleaseReceiveDataHandler(Topic topic)
        {
            this.receiveDataHandlerFactory.ReleaseReceiveDataHandler(topic, this);

            lock (partInfoData)
            {
                for (int i = 0; i < partInfoData.subscribeTopics.Count; i++)
                {
        			if (partInfoData.subscribeTopics[i].name.Equals(topic.GetName())) {
                        partInfoData.subscribeTopics.RemoveAt(i);
        				break;
		        	}
		        }
            }
        }

        // By modified singelton
        [MethodImpl(MethodImplOptions.Synchronized)]
        public ISendDataHandler GetSendDataHandler(Topic topic) 
        {
            return this.sendDataHandlerFactory.GetSendDataHandler(topic, this);
        }

        public Domain getDomain()
        {
            return this.domain;
        }

        public InProcessTransport GetInProcessTransport()
        {
            return this.inProcessTransport;
        }

        public void SaveConfig(string configFile)
        {
            if (this.config != null)
            {
                OPSConfig.SaveConfig(this.config, configFile);
            }
        }
    
        private void Run()
        {
            InitPartInfoData();

            while (this.doRun)
            {
                Thread.Sleep(1000);

                // Create publisher for participant info data
                if ( (partInfoPub == null) && (domain.GetMetaDataMcPort() != 0) )
                {
                    partInfoPub = new Publisher(CreateParticipantInfoTopic());
                }

                // Publish data
                if (partInfoPub != null)
                {
                    lock (partInfoData)
                    {
                        partInfoPub.WriteAsOPSObject(partInfoData);
                    }
                }
            }

            partInfoPub.Stop();
            partInfoPub = null;
        }

        private void SetupCyclicThread()
        {
            thread = new Thread(new ThreadStart(Run));
            thread.IsBackground = true;
            thread.Name = "ParticipantThread_" + domainID + "_" + participantID;
            thread.Start();
        }

    }

}